Content.Server/Xenoarchaeology/XenoArtifacts/ArtifactSystem.Nodes.cs

```diff
@@ -0,0 +1,198 @@
+        if (nodeAmount < 1)
+        {
+            Logger.Error($"nodeAmount {nodeAmount} is less than 1. Aborting artifact tree generation.");
+            return;
+        }
+
+        var uninitializedNodes = new List<ArtifactNode> { new() };
+        tree.StartNode = uninitializedNodes.First(); //the first node
+
+        while (uninitializedNodes.Any())
+        {
+            GenerateNode(ref uninitializedNodes, ref tree, nodeAmount);
+        }
+    }
+
+    /// <summary>
+    /// Generate an individual node on the tree.
+    /// </summary>
+    private void GenerateNode(ref List<ArtifactNode> uninitializedNodes, ref ArtifactTree tree, int targetNodeAmount)
+    {
+        if (!uninitializedNodes.Any())
+            return;
+
+        var node = uninitializedNodes.First();
+        uninitializedNodes.Remove(node);
+
+        node.Id = _random.Next(0, 10000);
+
+        //Generate the connected nodes
+        var maxEdges = Math.Max(1, targetNodeAmount - tree.AllNodes.Count - uninitializedNodes.Count - 1);
+        maxEdges = Math.Min(maxEdges, MaxEdgesPerNode);
+        var minEdges = Math.Clamp(targetNodeAmount - tree.AllNodes.Count - uninitializedNodes.Count - 1, 0, 1);
+
+        var edgeAmount = _random.Next(minEdges, maxEdges);
+
+        for (var i = 0; i < edgeAmount; i++)
+        {
+            var neighbor = new ArtifactNode
+            {
+                Depth = node.Depth + 1
+            };
+            node.Edges.Add(neighbor);
+            neighbor.Edges.Add(node);
+
+            uninitializedNodes.Add(neighbor);
+        }
+
+        node.Trigger = GetRandomTrigger(ref node);
+        node.Effect = GetRandomEffect(ref node);
+
+        tree.AllNodes.Add(node);
+    }
+
+    //yeah these two functions are near duplicates but i don't
+    //want to implement an interface or abstract parent
+
+    private ArtifactTriggerPrototype GetRandomTrigger(ref ArtifactNode node)
+    {
+        var allTriggers = _prototype.EnumeratePrototypes<ArtifactTriggerPrototype>().ToList();
+        var validDepth = allTriggers.Select(x => x.TargetDepth).Distinct().ToList();
+
+        var weights = GetDepthWeights(validDepth, node.Depth);
+        var selectedRandomTargetDepth = GetRandomTargetDepth(weights);
+        var targetTriggers = allTriggers.Where(x =>
+            x.TargetDepth == selectedRandomTargetDepth).ToList();
+
+        return _random.Pick(targetTriggers);
+    }
+
+    private ArtifactEffectPrototype GetRandomEffect(ref ArtifactNode node)
+    {
+        var allTriggers = _prototype.EnumeratePrototypes<ArtifactEffectPrototype>().ToList();
+        var validDepth = allTriggers.Select(x => x.TargetDepth).Distinct().ToList();
+
+        var weights = GetDepthWeights(validDepth, node.Depth);
+        var selectedRandomTargetDepth = GetRandomTargetDepth(weights);
+        var targetTriggers = allTriggers.Where(x =>
+            x.TargetDepth == selectedRandomTargetDepth).ToList();
+
+        return _random.Pick(targetTriggers);
```

Comment:
```suggestion
        var allEffects = _prototype.EnumeratePrototypes<ArtifactEffectPrototype>().ToList();
        var validDepth = allEffects.Select(x => x.TargetDepth).Distinct().ToList();

        var weights = GetDepthWeights(validDepth, node.Depth);
        var selectedRandomTargetDepth = GetRandomTargetDepth(weights);
        var targetEffects = allEffects.Where(x =>
            x.TargetDepth == selectedRandomTargetDepth).ToList();

        return _random.Pick(targetEffects);
```
confused me initially